File: cpp/3903.smallest-stable-index-i.cpp
#include <vector>
using std::vector;

class Solution {
public:
    int firstStableIndex(vector<int>& nums, int k) {
        const int n = nums.size();
        vector<int> vmax(n, nums.front());
        for (int i = 1; i < n; i++) {
            vmax[i] = std::max(vmax[i - 1], nums[i]);
        }

        vector<int> vmin(n, nums.back());
        for (int i = n - 2; i >= 0; i--) {
            vmin[i] = std::min(vmin[i + 1], nums[i]);
        }

        for (int i = 0; i < n; i++) {
            if (vmax[i] - vmin[i] <= k) {
                return i;
            }
        }

        return -1;
    }
};

File: rust/3903.smallest-stable-index-i.rs
impl Solution {
    pub fn first_stable_index(nums: Vec<i32>, k: i32) -> i32 {
        let n = nums.len();
        let mut vmax = vec![nums[0]; n];
        let mut vmin = vec![nums[n - 1]; n];

        for i in (1..n) {
          vmax[i] = vmax[i - 1].max(nums[i]);
        }

        for i in (0..n-1).rev() {
          vmin[i] = vmin[i + 1].min(nums[i]);
        }

        for i in (0..n) {
          if vmax[i] - vmin[i] <= k {
            return i as i32
          }
        }

        -1
    }
}